stefaniaPopescu26/assignment2/CommonAncestor.java

```diff
@@ -0,0 +1,227 @@
+                return right;
+
+            return null;
+        }
+
+        public void add (T parent_data, int direction, T data) {
+            /*
+                direction == 0 -> node to the left
+                direction == 1 -> node to the right
+             */
+            Node<T> aux;
+            Node<T> parent;
+
+            if (isEmpty()) {
+                root.setData(data);
+                return;
+            } else {
+                parent = findNode(root, parent_data);
+            }
+
+            if (parent == null)
+                return;
+
+            /*
+                If the parent already has a child to the left or to the right,
+                that child will be transferred to the new node.
+             */
+            if (direction == 0) {
+
+                if (parent.getLeft() != null) {
+                    aux = new Node<>(data, parent.getLeft(), null, parent);
+                } else {
+                    aux = new Node<>(data, null, null, parent);
+                }
+
+                parent.setLeft(aux);
+            } else {
+
+                if (parent.getRight() != null) {
+                    aux = new Node<>(data, null, parent.getRight(), parent);
+                } else {
+                    aux = new Node<>(data, null, null, parent);
+                }
+
+                parent.setRight(aux);
+            }
+        }
+
+        private boolean isCommonAncestor(Node<T> node, T data1, T data2) {
+            Node<T> left = node.getLeft();
+            Node<T> right = node.getRight();
+
+            if (left != null) {
+                if (node.getData() == data1)
+                    if (findNode(left, data2) != null)
+                        return true;
+                if (node.getData() == data2)
+                    if (findNode(left, data1) != null)
+                        return true;
+            }
+
+            if (right != null) {
+                if (node.getData() == data1)
+                    if (findNode(right, data2) != null)
+                        return true;
+                if (node.getData() == data2)
+                    if (findNode(right, data1) != null)
+                        return true;
+            }
+
+            if (left != null && right != null) {
+                if (findNode(left, data1) != null && findNode(right, data2) != null)
+                    return true;
+
+                if (findNode(left, data2) != null && findNode(right, data1) != null)
+                    return true;
+            }
+
+            return false;
+        }
+
+        private T LCA(Node<T> start, T data1, T data2) {
+            T LCAleft = null, LCAright = null;
+            Node<T> left = start.getLeft();
+            Node<T> right = start.getRight();
+
+            if (isCommonAncestor(start, data1, data2))
+                return start.getData();
+
+            if (left != null)
+                LCAleft = LCA(left, data1, data2);
+
+            if (right != null)
```

Comment:
These calls might not be necessary if the LCAleft is not null. You can move lines 186-187 before this to return early.